libff/algebra/curves/tests/test_groups.cpp

```diff
@@ -26,6 +25,22 @@
 
 using namespace libff;
 
+// Utility function to compute a point on the curve E(Fq) with the given x
+// coordinate.
+template<typename GroupT>
+GroupT curve_point_at_x(const typename GroupT::base_field &x)
+{
+    const typename GroupT::base_field x_squared = x * x;
+    const typename GroupT::base_field x_cubed = x_squared * x;
+    const typename GroupT::base_field y_squared =
+        x_cubed + (GroupT::coeff_a * x_squared) + GroupT::coeff_b;
+    // Assert that y_squared is in a quadatic residue (ensurign that sqrt()
+    // terminates).
+    assert((y_squared^GroupT::base_field::euler) == GroupT::base_field::one());
```

Comment:
Hum.. I am not sure an `assert` is the right way of checking Quad. Res. here. Won't the `assert` be ignored if the program runs in non-debug mode?
Since I can see that a similar function is used here: https://github.com/clearmatics/zeth/pull/327/files#r544447282 wouldn't that make sense to move this "test util function" to the library and move the QR check outside of the assert (returning a default value if the the RHS of the curve equation does not yield a QR (or throwing an exception))?

Reply:
> Won't the assert be ignored if the program runs in non-debug mode?

This makes me think that we really need to switch to GTest... Upstream has switched to that already. That'd be good to align with it on that note..

Reply:
> I am not sure an assert is the right way of checking Quad. Res. here. Won't the assert be ignored if the program runs in non-debug mode?
> ...
> This makes me think that we really need to switch to GTest... Upstream has switched to that already. That'd be good to align with it on that note..

The unit tests all rely on forcing asserts on.  I completely agree that gtest would be better.  (#11).

> Since I can see that a similar function is used here: https://github.com/clearmatics/zeth/pull/327/files#r544447282 wouldn't that make sense to move this "test util function" to the library and move the QR check outside of the assert (returning a default value if the the RHS of the curve equation does not yield a QR (or throwing an exception))?

We could add a util function somewhere.  In this particular case, I have actually checked in advance that the points we generate are on the curve, so it might be better to address this separately.

Reply:
> We could add a util function somewhere. In this particular case, I have actually checked in advance that the points we generate are on the curve, so it might be better to address this separately.

Actually, a `bool fp_model::is_quaratic_residue()` method would be trivial to add and could be pretty helpful.  (Is that what you are suggesting?)

Reply:
I've added `curve_util` functions to find these points a2f9badd3a8b900f5b30d4313f96bebe5fb05c63.  For the twisted curve, without some common method over all extension fields to check that sqrt terminates (which might be a large enough change to do in a separate PR) we are unable to make the same check.

Let me know what you think.

Reply:
> Actually, a bool fp_model::is_quaratic_residue() method would be trivial to add and could be pretty helpful. (Is that what you are suggesting?)
> (EDIT: I guess you are referring to the whole function to compute the point)

Yes I was referring to the whole function (as you mention in your "EDIT"). That being said, having a function `bool fp_model::is_quaratic_residue()` is also a good idea!

Reply:
>  I've added curve_util functions to find these points a2f9bad. For the twisted curve, without some common method over all extension fields to check that sqrt terminates (which might be a large enough change to do in a separate PR) we are unable to make the same check.
> Let me know what you think.

Great thanks. I think this is a good place to have this function.